libbpfgo: Add map batch operations

derekparker opened the pull request:
This patch adds a few batch operations for maps including:

Lookup, LookupAndDelete, Update and Delete.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aquasecurity/libbpfgo?pullRequest=97) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/libbpfgo?pullRequest=97) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/libbpfgo?pullRequest=97) it.</sub>

rafaeldtinoco commented on Makefile:
```diff
@@ -75,13 +75,14 @@ libbpfgo-static: $(VMLINUXH) | $(LIBBPF_OBJ)
-		sudo -E -- go test \
-		-tags netgo -ldflags $(CGO_EXTLDFLAGS_STATIC) \
-		.
+	CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
```
could you, pls, keep the indentation here as it was ? I'm going to "fix" the Makefile style soon, not only for this line but for other lines as well.

derekparker replied:
Ack, no problem!

rafaeldtinoco reviewed (dismissed):
LGTM. We should open an issue to transform the "selftest/build" into a proper selftest/XXX directory. @grantseltzer WDYT ?

Besides, I think we should "unify" the Makefiles into a single Makefile at the root dir (just like we recently did to tracee). 

Nevertheless, I think we should accept this contribution, which was great.

@grantseltzer I'll let you squash & merge this if you agree.

grantseltzer commented on libbpfgo.go:
```diff
@@ -578,6 +578,116 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
+		Flags:     C.BPF_ANY,
+	}
+
+	errC := C.bpf_map_lookup_batch(b.fd, startKey, nextKey, keys, valuesPtr, &countC, bpfMapBatchOptsToC(opts))
+	if errC != 0 {
+		return nil, fmt.Errorf("failed to batch lookup values %v in map %s: %d", keys, b.name, errC)
+	}
```
errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

derekparker replied:
Apologies, I followed the example of the other BPFMap methods (such as `Pin`, `UnPin`, etc...) which don't do any extra checking on the returned error than what I've done. Do you have another specific example or pattern you'd like me to follow or extra checks to perform?

grantseltzer replied:
I had an answer ready to this but then after carefully examining libbpf code I got confused. I asked on the mailing list how functions which use `libbpf_err` vs `libbpf_err_errno` are supposed to be checked, let's wait for a response from Andrii before proceededing:

https://lore.kernel.org/bpf/CAO658oUqd5=B3zkDhm2jVQxG+vEf=2CE7WimXHqgcH+m0P=k_Q@mail.gmail.com/T/#u

I'm sure there's functions in libbpfgo that aren't correct.

grantseltzer replied:
@yanivagman You probably have a good idea

derekparker replied:
Ack, sounds good to me!

derekparker replied:
If I understand correctly it looks like Andrii responded saying that the returned value can be used directly without checking errno unless it's an API which returns a pointer, in which case errno must be checked.

Are you reading the response the same?

grantseltzer replied:
Yes I read it the same, which means the way you're doing it is correct. Though perhaps we can extract the error code to determine what the issue was. I think that's what using the `syscall.Errno` type does, but I will investigate in the morning.

grantseltzer commented on libbpfgo.go:
```diff
@@ -578,6 +578,116 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
+		Flags:     C.BPF_ANY,
+	}
+
+	errC := C.bpf_map_lookup_and_delete_batch(b.fd, startKey, nextKey, keys, valuesPtr, &countC, bpfMapBatchOptsToC(opts))
+	if errC != 0 {
+		return nil, fmt.Errorf("failed to batch lookup and delete values %v in map %s", keys, b.name)
+	}
```
errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

grantseltzer commented on libbpfgo.go:
```diff
@@ -578,6 +578,116 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
+		ElemFlags: C.BPF_ANY,
+		Flags:     C.BPF_ANY,
+	}
+	errC := C.bpf_map_delete_batch(b.fd, keys, &countC, bpfMapBatchOptsToC(opts))
+	if errC != 0 {
+		return fmt.Errorf("failed to get lookup key %d from map %s", keys, b.name)
+	}
```
errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

grantseltzer commented on libbpfgo.go:
```diff
@@ -578,6 +578,116 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
+		ElemFlags: C.BPF_ANY,
+		Flags:     C.BPF_ANY,
+	}
+	errC := C.bpf_map_update_batch(b.fd, keys, values, &countC, bpfMapBatchOptsToC(&opts))
+	if errC != 0 {
+		return fmt.Errorf("failed to update map %s: %v", b.name, errC)
+	}
+	return nil
```
errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

grantseltzer commented on Makefile:
```diff
@@ -75,11 +75,12 @@ libbpfgo-static: $(VMLINUXH) | $(LIBBPF_OBJ)
 		GOOS=linux GOARCH=$(ARCH) \
-		sudo -E -- go test \
+		go test \
 		-tags netgo -ldflags $(CGO_EXTLDFLAGS_STATIC) \
```
can you add `-v` in here while you're at it?

derekparker replied:
Done.

grantseltzer reviewed (changes requested):
Thanks so much for this contribution Derek! 

In addition to my inline comments, can I ask that you move your tests into `selftests`? It shouldn't make much of a change, just copy the directory structure of one of the other selftests.

I'm not entirely sure why we still have the test that's in `libbpfgo_test.go` (Test_LoadAndAttach), that file should contain tests that don't shell out, things like helper functions and whatnot. I'm going to create an issue to track removing/replacing that one.

derekparker commented on libbpfgo.go:
```diff
@@ -578,6 +578,116 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
+
+// GetValueBatch allows for batch lookups of multiple keys.
+// The first argument is a pointer to an array of keys which will be populated with the keys returned from this operation.
+// It returns the associated values as a slice of slices of bytes.
```
That SGTM, will update.

derekparker replied:
Actually, the way I would implement this would be to return `map[[]byte][]byte` because we don't know what type the keys will be (or size). The problem with this is a `[]byte` is not comparable so cannot be used as a key. I suppose we could use a `map[interface{}][]byte` and that should work (haven't tested it) but feels kind of weird to me, and adds a type assertion to the callers code.

I'm happy to implement the latter if that's what you'd prefer, however this interface matches what happens in libbpf (although admittedly it's not very Go-like).

As shown in the test it's fairly straightforward still to iterate through keys / values using a single iterator var to keep track of where you are.

grantseltzer replied:
Good points, best to leave it as is.

derekparker replied:
Sounds good!

derekparker commented:
Hey @grantseltzer thanks for the feedback!

I've addressed most of your review except for that which I've commented on. Also I've moved the test over to it's own directory in `selftest/map-batch` and removed the corresponding tests in `libbpfgo_test.go`.

grantseltzer commented on selftest/map-batch/main.go:
```diff
@@ -0,0 +1,128 @@
+	prevKey := unsafe.Pointer(nil)
+	step := len(batchKeys)
+	for i := 0; i < 2; i++ {
+		if i > 0 {
+			// We're on step 2, so test the batch lookup by specifying the
+			// key to start from.
+			prevKey = unsafe.Pointer(&nextKey)
+		}
```
Can you explain more of what's going on here? Why is there two steps? I see the documentation for BPF_MAP_LOOKUP_BATCH in uapi/linux/bpf.h but I don't fully understand why it's designed this way.

derekparker replied:
Yeah, so the API is such that you can iterate over all the keys/values in batches. Essentially you provide a pointer to the last key seen, the next key to start with, and a buffer for keys/values.

That way if you have a map with max size of 1000 for example you can iterate over the entire thing in chunks instead of loading everything into memory. It's kind of a confusing API to be honest it took me a bit to wrap my head around it when implementing.

grantseltzer replied:
Got it - that took some time to understand. Definitely worth documenting upstream in libbpf.

grantseltzer replied:
After that this is good to go :+1:

derekparker replied:
> Got it - that took some time to understand. Definitely worth documenting upstream in libbpf.

Yes! I literally had to read through the test code to understand how to use the API.

derekparker replied:
Ok, let me know if the documentation I added is clear enough.

derekparker commented on selftest/map-batch/main.bpf.c:
```diff
@@ -0,0 +1,43 @@
+    __uint(key_size, sizeof(u32));
+    __uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
```
Removed.

derekparker commented on selftest/map-batch/main.bpf.c:
```diff
@@ -0,0 +1,43 @@
+struct value {
+    int x;
+    char y;
+};
```
Done, forgot to do that when swapping over.

derekparker commented on selftest/map-batch/main.bpf.c:
```diff
@@ -0,0 +1,43 @@
+    bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, v2, sizeof(char)*3);
+
+    return 0;
+}
```
Done.

derekparker commented on selftest/map-batch/main.go:
```diff
@@ -0,0 +1,128 @@
+			// key to start from.
+			prevKey = unsafe.Pointer(&nextKey)
+		}
+		vals, err := testerMap.GetValueBatch(unsafe.Pointer(&batchKeys[0]), prevKey, unsafe.Pointer(&nextKey), uint32(step))
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "step %d testerMap.LookupBatch failed: %v", i, err)
+			os.Exit(-1)
+		}
```
Ack, that's my bad forgot to update when swapping over from the other test.

derekparker commented on selftest/map-batch/main.go:
```diff
@@ -0,0 +1,128 @@
+	if err = resizeMap(bpfModule, "events", 8192); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
```
Removed.

derekparker commented on selftest/map-batch/main.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	return nil
+}
```
Removed.

grantseltzer commented on libbpfgo.go:
```diff
@@ -578,6 +578,120 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
+// This API allows for batch lookups of multiple keys, potentially in steps over multiple iterations. For example,
+// you provide the last key seen (or nil) for the startKey, and the first key to start the next iteration with in nextKey.
+// Once the first iteration is complete you can provide the last key seen in the previous iteration as the startKey for the next iteration
+// and for forth until nextKey is nil.
```
```suggestion
// and repeat until nextKey is nil.
```

grantseltzer commented:
Made a small edit I didn't want to bother you with. This looks good to me! Thanks so much @derekparker !
